src/execution/exec/execution_context.cpp

```diff
@@ -67,78 +99,89 @@ void ExecutionContext::EndPipelineTracker(query_id_t query_id, pipeline_id_t pip
-            feature.SetCardinality(value);
-            recorded = true;
-            break;
-          }
-          case brain::ExecutionOperatingUnitFeatureAttribute::NUM_LOOPS: {
-            feature.SetNumLoops(value);
-            recorded = true;
-            break;
-          }
-          default:
-            UNREACHABLE("Invalid feature attribute.");
-        }
-        break;
-      }
+void ExecutionContext::InitializeParallelOUFeatureVector(brain::ExecOUFeatureVector *ouvec, pipeline_id_t pipeline_id) {
+  auto *vec = new (ouvec) brain::ExecOUFeatureVector();
+  vec->pipeline_id_ = pipeline_id;
+  vec->pipeline_features_ =
+      std::make_unique<execution::sql::MemPoolVector<brain::ExecutionOperatingUnitFeature>>(GetMemoryPool());
+
+  bool found_blocking = false;
+  brain::ExecutionOperatingUnitFeature feature;
+  auto features = pipeline_operating_units_->GetPipelineFeatures(pipeline_id);
+  for (auto &feat : features) {
+    if (brain::OperatingUnitUtil::IsOperatingUnitTypeBlocking(feat.GetExecutionOperatingUnitType())) {
+      TERRIER_ASSERT(!found_blocking, "Pipeline should only have 1 blocking");
+      found_blocking = true;
+      feature = feat;
     }
-    TERRIER_ASSERT(recorded, "Nothing was recorded. OperatingUnitRecorder hacks are probably necessary.");
+  }
+
+  if (!found_blocking) {
+    TERRIER_ASSERT(false, "Pipeline should have 1 blocking");
+    return;
+  }
+
+  switch (feature.GetExecutionOperatingUnitType()) {
+    case brain::ExecutionOperatingUnitType::HASHJOIN_BUILD:
+      vec->pipeline_features_->emplace_back(brain::ExecutionOperatingUnitType::PARALLEL_MERGE_HASHJOIN, feature);
+      break;
+    case brain::ExecutionOperatingUnitType::AGGREGATE_BUILD:
+      vec->pipeline_features_->emplace_back(brain::ExecutionOperatingUnitType::PARALLEL_MERGE_AGGBUILD, feature);
+      break;
+    case brain::ExecutionOperatingUnitType::SORT_BUILD:
+      vec->pipeline_features_->emplace_back(brain::ExecutionOperatingUnitType::PARALLEL_SORT_STEP, feature);
+      vec->pipeline_features_->emplace_back(brain::ExecutionOperatingUnitType::PARALLEL_SORT_MERGE_STEP, feature);
+      break;
+    case brain::ExecutionOperatingUnitType::CREATE_INDEX:
+      vec->pipeline_features_->emplace_back(brain::ExecutionOperatingUnitType::CREATE_INDEX_MAIN, feature);
+      break;
+    default:
+      TERRIER_ASSERT(false, "Unsupported parallel OU");
+  }
+
+  // Update num_concurrent
+  for (auto &feature : *vec->pipeline_features_) {
+    feature.SetNumConcurrent(num_concurrent_estimate_);
   }
 }
 
 const parser::ConstantValueExpression &ExecutionContext::GetParam(const uint32_t param_idx) const {
   return (*params_)[param_idx];
 }
 
+void ExecutionContext::RegisterHook(size_t hook_idx, HookFn hook) {
+  TERRIER_ASSERT(hook_idx < hooks_.capacity(), "Incorrect number of reserved hooks");
+  hooks_[hook_idx] = hook;
+}
+
+void ExecutionContext::InvokeHook(size_t hookIndex, void *tls, void *arg) {
```

Comment:
snake_pls notCamel

I thought the linter would catch this, actually. @lmwnshn check